Add register_autograd to register backward formulas for custom ops

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123454
* #123453
* #123437
* #123129
* __->__ #123110
* #123109
* #123108

The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

zou3519 committed:
Add register_autograd to register backward formulas for custom ops

The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -275,6 +277,78 @@ def register_fake(self, fn: Callable) -> Callable:
+            >>>     x, = ctx.saved_tensors
+            >>>     return grad * x.cos()
+            >>>
+            >>> numpy_sin.register_autograd(setup_context, backward)
```
Taking API bikeshedding feedback

albanD replied:
One option that comes to mind since setup_context is optional is to have this decorator be around backward_fn and take in only setup_context?

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

albanD commented on torch/_library/autograd.py:
```diff
@@ -0,0 +1,37 @@
+
+def make_autograd_impl(opdef: Any) -> Callable:
+    def autograd_impl(*args):
+        name: str = "GeneratedBackwardFor" + opdef._name
```
I think you want to use both namespace and name here?
Also CamelCase them?

zou3519 replied:
> I think you want to use both namespace and name here?

yup

> Also CamelCase them?

This will make it more difficult to read

albanD commented on torch/_library/autograd.py:
```diff
@@ -0,0 +1,37 @@
+
+        def forward(ctx, *args):
+            with _C._AutoDispatchBelowAutograd():
+                result = opdef._opoverload(*args)
```
This should really be a redispatch...

zou3519 replied:
Hmm let me see how difficult it is to add a redispatch to the Python api

zou3519 replied:
This seems like it is going to be difficult. Also, what exactly does this change?

I don't think that changing this from a dispatch to a redispatch later will be too BC-breaking (because we can't register an implementation in Python that redispatches before/after and users generally don't know what a redispatch is).

zou3519 replied:
Alban:
- how does this interact with backendselect? (there are some subtle interactions).
- how does it interact with things "around the dispatcher", aka the profiler. You'll get two events in the profiler!
- also, the TLSSnapshot thing is unclear(?) but probably OK.

Alban: OK with shipping this first as long as we do the redispatch next, before this thing goes fully public. Blocker for wider adoption (TODO(rzou): file an issue)

albanD commented on torch/_library/autograd.py:
```diff
@@ -0,0 +1,37 @@
+                f"Please use register_autograd to add one."
+            )
+
+        Generated = type(
```
nit: you can generate this class in the make_autograd_impl context and only do the apply() in the autograd_impl no?

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -277,6 +279,78 @@ def register_fake(self, fn: Callable) -> Callable:
+
+        Both ``setup_context_fn`` and ``backward_fn`` must be traceable. That is,
+        they may not directly access Tensor.data_ptr and they must not depend on
+        or mutate global state.
```
```suggestion
        or mutate global state. If you need a non-traceable backward, you can make it
        a separate custom_op that you call inside ``backward_fn``.
```

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -277,6 +279,78 @@ def register_fake(self, fn: Callable) -> Callable:
+            >>>     y_np = np.sin(x_np)
+            >>>     return torch.from_numpy(y_np).to(device=x.device)
+            >>>
+            >>> def setup_context(ctx, inputs, output) -> Tensor:
```
I guess this is `outputs`?

zou3519 replied:
No, this is just output. If the output is a single Tensor, we do not promote it to a Tuple.

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

albanD commented on torch/_library/autograd.py:
```diff
@@ -0,0 +1,37 @@
+        if opdef._backward_fn:
+            return opdef._backward_fn(ctx, *grads)
+        raise RuntimeError(
+            f"There was no autograd formula registered for {opdef}. "
+            f"Please use register_autograd to add one."
```
```suggestion
            f"Trying to backward through {opdef} but no "
            "autograd formula has been registered. "
            "Please use register_autograd to add one."
```

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -283,6 +285,79 @@ def register_fake(self, fn: Callable) -> Callable:
+        2. You must tell us how to compute gradients during the backward pass.
+           This is the "backward" function.
+
+        ``setup_context_fn(ctx, inputs, output)`` runs during the forward pass.
```
Yup

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]

zou3519 committed:
Update on "Add register_autograd to register backward formulas for custom ops"


The user provides a `setup_context` and a `backward_function`. These
get put into a torch.autograd.Function that gets registered as the
custom op's autograd implementation.

Test Plan:
- we update custom ops in the custom_op_db to use the new
  register_autograd API.

[ghstack-poisoned]
